test/extensions/config_subscription/grpc/xds_failover_integration_test.cc

```diff
@@ -906,4 +906,149 @@ TEST_P(XdsFailoverAdsIntegrationTest,
+  config_helper_.addConfigModifier([](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+    auto* lds_config = bootstrap.mutable_dynamic_resources()->mutable_lds_config();
+    lds_config->mutable_initial_fetch_timeout()->set_seconds(100);
+  });
+  initialize();
+
+  // 2 consecutive primary failures.
+  // Expect a connection to the primary. Reject the connection immediately.
+  primaryConnectionFailure();
+  ASSERT_TRUE(xds_connection_->waitForDisconnect());
+  // The CDS request fails when the primary disconnects. After that fetch the config
+  // dump to ensure that the retry timer kicks in.
+  // Expect another connection attempt to the primary. Reject the stream (gRPC failure) immediately.
+  // As this is a 2nd consecutive failure, it will trigger failover.
+  waitForPrimaryXdsRetryTimer();
+  primaryConnectionFailure();
+  ASSERT_TRUE(xds_connection_->waitForDisconnect());
+
+  // The CDS request fails when the primary disconnects.
+  test_server_->waitForCounterGe("cluster_manager.cds.update_failure", 2);
+
+  AssertionResult result =
+      failover_xds_upstream_->waitForHttpConnection(*dispatcher_, failover_xds_connection_);
+  RELEASE_ASSERT(result, result.message());
+  // Failover is healthy, start the ADS gRPC stream.
+  result = failover_xds_connection_->waitForNewStream(*dispatcher_, failover_xds_stream_);
+  RELEASE_ASSERT(result, result.message());
+  failover_xds_stream_->startGrpcStream();
+
+  // Ensure basic flow with failover works.
+  EXPECT_TRUE(compareDiscoveryRequest(CdsTypeUrl, "", {}, {}, {}, true,
+                                      Grpc::Status::WellKnownGrpcStatus::Ok, "",
+                                      failover_xds_stream_.get()));
+  sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+      CdsTypeUrl, {ConfigHelper::buildCluster("failover_cluster_0")},
+      {ConfigHelper::buildCluster("failover_cluster_0")}, {}, "failover1", {},
+      failover_xds_stream_.get());
+  // Wait for an EDS request, and send its response.
+  test_server_->waitForGaugeEq("cluster_manager.warming_clusters", 1);
+  test_server_->waitForGaugeEq("cluster.failover_cluster_0.warming_state", 1);
+  // Ensure basic flow with failover works.
+  EXPECT_TRUE(compareDiscoveryRequest(
+      EdsTypeUrl, "", {"failover_cluster_0"}, {"failover_cluster_0"}, {}, false,
+      Grpc::Status::WellKnownGrpcStatus::Ok, "", failover_xds_stream_.get()));
+  sendDiscoveryResponse<envoy::config::endpoint::v3::ClusterLoadAssignment>(
+      EdsTypeUrl, {buildClusterLoadAssignment("failover_cluster_0")},
+      {buildClusterLoadAssignment("failover_cluster_0")}, {}, "failover1", {},
+      failover_xds_stream_.get());
+  test_server_->waitForGaugeEq("cluster_manager.warming_clusters", 0);
+  test_server_->waitForGaugeGe("cluster_manager.active_clusters", 2);
+  test_server_->waitForGaugeEq("cluster.failover_cluster_0.warming_state", 0);
+  EXPECT_EQ(2, test_server_->gauge("control_plane.connected_state")->value());
```

Comment:
was always thinking that this stat can only have either `0` or `1` value.

Reply:
This was true until #35691. We then changed it to: 0 - not connected, 1 - connected to the primary, 2 - connected to the failover.

Reply:
Should we update the docs? https://www.envoyproxy.io/docs/envoy/latest/configuration/overview/mgmt_server#statistics

Reply:
The docs will be updated once we get to thoroughly test the xDS-failover paths and gain confidence that it works as expected. (see #35746 for the relevant doc changes, including stats doc update)

Reply:
aha! got it